pkg/storage/ingest/writer.go

```diff
@@ -189,38 +194,52 @@ func (w *Writer) WriteSync(ctx context.Context, partitionID int32, userID string
 // or an error occurred.
 func (w *Writer) produceSync(ctx context.Context, client *kgo.Client, records []*kgo.Record) kgo.ProduceResults {
 	var (
-		remaining = atomic.NewInt64(int64(len(records)))
-		done      = make(chan struct{})
-		resMx     sync.Mutex
-		res       kgo.ProduceResults
+		remaining        = atomic.NewInt64(int64(len(records)))
+		done             = make(chan struct{})
+		resMx            sync.Mutex
+		res              = make(kgo.ProduceResults, 0, len(records))
+		maxBufferedBytes = int64(w.kafkaCfg.ProducerMaxBufferedBytes)
 	)
 
 	w.writeRequestsTotal.Add(float64(len(records)))
 
+	onProduceDone := func(r *kgo.Record, err error) {
+		if maxBufferedBytes > 0 {
+			w.writersBufferedBytes.Add(-int64(len(r.Value)))
+		}
+
+		resMx.Lock()
```

Comment:
nit: this lock would not be necessary if we'd pre-allocate `res` with the `len()` which is `len(records)` and in the record's context (`kgo.Record.Context`) we'd set the record's index, then in `onProduceDone` we could just assign to `res[recordIndex] = <result>` without lock.

Reply:
nvm, adding a property like this to the context would probably create an allocation, that's not worth it...